lib/vrl/vrl/benches/vm.rs

```diff
@@ -0,0 +1,67 @@
+use std::collections::BTreeMap;
+
+use compiler::state;
+use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
+use indoc::indoc;
+use vector_common::TimeZone;
+use vrl::{Runtime, Value};
+
+static SOURCE: [(&str, &str); 2] = [
+    (
+        "parse_json",
+        indoc! {r#"
+            x = parse_json!(s'{"noog": "nork"}')
+            x.noog
+        "#},
+    ),
+    (
+        "simple",
+        indoc! {r#"
+            .hostname = "vector"
+
+            if .status == "warning" {
+            .thing = upcase(.hostname)
+            } else if .status == "notice" {
+            .thung = downcase(.hostname)
+            } else {
+            .nong = upcase(.hostname)
+            }
+
+            .matches = { "name": .message, "num": "2" }
+            .origin, .err = .hostname + "/" + .matches.name + "/" + .matches.num
+        "#},
+    ),
+];
+
+fn benchmark_kind_display(c: &mut Criterion) {
+    let mut group = c.benchmark_group("vrl_compiler/value::kind::display");
+    for (name, source) in &SOURCE {
+        let state = state::Runtime::default();
+        let runtime = Runtime::new(state);
+        let tz = TimeZone::default();
+        let functions = vrl_stdlib::all();
+        let program = vrl::compile(source, &functions, None).unwrap();
+        let vm = runtime.compile(functions, &program).unwrap();
+
+        group.bench_with_input(BenchmarkId::new("Vm", name), &vm, |b, vm| {
+            b.iter(|| {
+                let state = state::Runtime::default();
+                let mut runtime = Runtime::new(state);
+                runtime.run_vm(vm, &mut Value::Object(BTreeMap::default()), &tz)
+            })
+        });
+
+        group.bench_with_input(BenchmarkId::new("Ast", name), &(), |b, _| {
+            b.iter(|| {
+                let state = state::Runtime::default();
+                let mut runtime = Runtime::new(state);
+                runtime.resolve(&mut Value::Object(BTreeMap::default()), &program, &tz)
```

Comment:
I think we should rely on https://docs.rs/criterion/latest/criterion/fn.black_box.html here to avoid this potentially being completely optimized out.

E.g.

```suggestion
                let mut target = Value::Object(BTreeMap::default());
                runtime.resolve(&mut target, &program, &tz);
                criterion::black_box(target);
```

Reply:
Yeah, it couldn't hurt, though I think returning the `Result` from `runtime.resolve` value to the closure, as it is here, should also avoid the compiler optimizing it out the call.

Reply:
Yeah, TBH, what is and is not going to be optimized out shifts around so much that it's just one of those dings against the effectiveness of micro-benchmarks.